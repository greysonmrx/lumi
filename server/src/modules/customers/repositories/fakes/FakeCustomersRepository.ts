import { randomUUID } from "node:crypto";

import { ICreateCustomerDTO } from "../../dtos/ICreateCustomerDTO";
import { IFindCustomerByNumberDTO } from "../../dtos/IFindCustomerByNumberDTO";

import { Customer } from "../../entities/fakes/Customer";

import { ICustomersRepository } from "../ICustomersRepository";

export class FakeCustomersRepository implements ICustomersRepository {
  private customers: Customer[] = [];

  public async findByNumber(
    data: IFindCustomerByNumberDTO
  ): Promise<Customer | null> {
    const findCustomer = this.customers.find(
      (customer) => customer.number === data.number
    );

    return findCustomer ?? null;
  }

  public async create(data: ICreateCustomerDTO): Promise<Customer> {
    const customer = new Customer();

    Object.assign(customer, {
      id: randomUUID(),
      name: data.name,
      number: data.number,
      createdAt: new Date(),
    });

    this.customers.push(customer);

    return customer;
  }
}
