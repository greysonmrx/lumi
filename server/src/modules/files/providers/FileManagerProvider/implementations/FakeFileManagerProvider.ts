import { generateFileKey } from "@/shared/utils/generateFileKey";

import { IFileManagerProvider } from "../models/IFileManagerProvider";
import { IFileToBeUploaded } from "../models/IFileToBeUploaded";

export class FakeFileManagerProvider implements IFileManagerProvider {
  public prepareUpload(file: IFileToBeUploaded): Pick<
    IFileToBeUploaded,
    "name"
  > & {
    path: string;
    upload: () => Promise<void>;
  } {
    const name = generateFileKey(file.name);
    const path = "/fake/path";

    return {
      name,
      path,
      upload: async () => {},
    };
  }
}
