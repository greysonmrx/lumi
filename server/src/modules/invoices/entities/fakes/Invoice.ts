import { IFile } from "@/modules/files/entities/IFile";
import { ICustomer } from "@/modules/customers/entities/ICustomer";

import { IInvoice } from "../IInvoice";
import { IInvoiceItem } from "../IInvoiceItem";

export class Invoice implements IInvoice {
  public id: string;
  public dueDate: Date;
  public items?: IInvoiceItem[];
  public customer?: ICustomer;
  public customerId: string;
  public pdf?: IFile;
  public pdfId?: string;
  public referenceMonth: string;
  public totalValue: number;
  public createdAt: Date;
}
