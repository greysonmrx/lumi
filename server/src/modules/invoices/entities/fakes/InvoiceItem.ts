import { IInvoiceItem } from "../IInvoiceItem";

export class InvoiceItem implements IInvoiceItem {
  public id: string;
  public name: string;
  public amount?: number;
  public invoiceId: string;
  public totalValue: number;
  public unitPrice?: number;
}
