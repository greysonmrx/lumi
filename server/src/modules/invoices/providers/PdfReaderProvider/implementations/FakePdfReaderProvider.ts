import { IPdfReaderProvider } from "../models/IPdfReaderProvider";

export class FakePdfReaderProvider implements IPdfReaderProvider {
  public async getContent(pdfFile: Buffer): Promise<string> {
    return String(pdfFile);
  }
}
