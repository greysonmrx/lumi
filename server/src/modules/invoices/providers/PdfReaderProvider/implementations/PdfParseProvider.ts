import PdfParse from "pdf-parse";

import { IPdfReaderProvider } from "../models/IPdfReaderProvider";

export class PdfParseProvider implements IPdfReaderProvider {
  public async getContent(pdfFile: Buffer): Promise<string> {
    const pdfData = await PdfParse(pdfFile);

    return pdfData.text;
  }
}
