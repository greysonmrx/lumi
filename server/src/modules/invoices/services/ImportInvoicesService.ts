import { NotFoundError } from "@/shared/errors/NotFoundError";

import { INVOICE_ITEMS_TO_FIND } from "@/config/constants";

import { IInvoicesRepository } from "../repositories/IInvoicesRepository";
import { IFilesRepository } from "@/modules/files/repositories/IFilesRepository";
import { ICustomersRepository } from "@/modules/customers/repositories/ICustomersRepository";

import { IFileToBeUploaded } from "@/modules/files/providers/FileManagerProvider/models/IFileToBeUploaded";
import { IFileManagerProvider } from "@/modules/files/providers/FileManagerProvider/models/IFileManagerProvider";
import { IPdfReaderProvider } from "../providers/PdfReaderProvider/models/IPdfReaderProvider";

import { IInvoice } from "../entities/IInvoice";

import { getDueDate } from "../helpers/getDueDate";
import { getTotalValue } from "../helpers/getTotalValue";
import { getCustomerName } from "../helpers/getCustomerName";
import { getCustomerNumber } from "../helpers/getCustomerNumber";
import { getReferenceMonth } from "../helpers/getReferenceMonth";
import {
  getInvoiceItemData,
  type InvoiceItemData,
} from "../helpers/getInvoiceItemData";

type ImportInvoicesServiceRequest = {
  files: IFileToBeUploaded[];
};

type ImportInvoicesServiceResponse = IInvoice[];

export class ImportInvoicesService {
  constructor(
    private readonly filesRepository: IFilesRepository,
    private readonly invoicesRepository: IInvoicesRepository,
    private readonly customersRepository: ICustomersRepository,
    private readonly pdfReaderProvider: IPdfReaderProvider,
    private readonly fileManagerProvider: IFileManagerProvider
  ) {}

  public async perform({
    files,
  }: ImportInvoicesServiceRequest): Promise<ImportInvoicesServiceResponse> {
    const invoices: IInvoice[] = [];

    for (const file of files) {
      const pdfFile = this.fileManagerProvider.prepareUpload(file);

      const pdfContent = await this.pdfReaderProvider.getContent(file.content);

      const pdfContentInLines = pdfContent
        .replace(/ {1,}/g, " ")
        .split("\n")
        .map((line) => line.trim());

      const customerName = this.getCustomerName(pdfContentInLines);

      if (!customerName) {
        throw new NotFoundError("Nome do cliente não foi encontrado");
      }

      const customerNumber = this.getCustomerNumber(pdfContentInLines);

      if (!customerNumber) {
        throw new NotFoundError("Número do cliente não foi encontrado");
      }

      let customer = await this.customersRepository.findByNumber({
        number: customerNumber,
      });

      const dueDate = this.getDueDate(pdfContentInLines);

      if (!dueDate) {
        throw new NotFoundError("Data de vencimento não foi encontrada");
      }

      const referenceMonth = this.getReferenceMonth(pdfContentInLines);

      if (!referenceMonth) {
        throw new NotFoundError("Mês de referência não foi encontrado");
      }

      const totalValue = this.getTotalValue(pdfContentInLines);

      if (!totalValue) {
        throw new NotFoundError("Valor total não foi encontrado");
      }

      const invoiceItems: InvoiceItemData[] = [];

      Object.keys(INVOICE_ITEMS_TO_FIND).forEach((INVOICE_ITEM) => {
        const invoiceItem = this.getInvoiceItemData(
          pdfContentInLines,
          INVOICE_ITEM as keyof typeof INVOICE_ITEMS_TO_FIND
        );

        if (invoiceItem) {
          invoiceItems.push(invoiceItem);
        }
      });

      if (!customer) {
        customer = await this.customersRepository.create({
          name: customerName,
          number: customerNumber,
        });
      }

      const { upload: uploadPdf, ...pdfData } = pdfFile;

      const pdf = await this.filesRepository.create(pdfData);

      const [day, month, year] = dueDate.split("/");

      const invoice = await this.invoicesRepository.create({
        pdfId: pdf.id,
        items: invoiceItems,
        dueDate: new Date(`${year}-${month}-${day}`),
        totalValue,
        referenceMonth,
        customerId: customer.id,
      });

      await uploadPdf();

      invoices.push(invoice);
    }

    return invoices;
  }

  public getCustomerName(contentInLines: string[]) {
    return getCustomerName(contentInLines);
  }

  public getDueDate(contentInLines: string[]) {
    return getDueDate(contentInLines);
  }

  public getTotalValue(contentInLines: string[]) {
    return getTotalValue(contentInLines);
  }

  public getCustomerNumber(contentInLines: string[]) {
    return getCustomerNumber(contentInLines);
  }

  public getReferenceMonth(contentInLines: string[]) {
    return getReferenceMonth(contentInLines);
  }

  public getInvoiceItemData(
    contentInLines: string[],
    itemName: keyof typeof INVOICE_ITEMS_TO_FIND
  ) {
    return getInvoiceItemData(contentInLines, itemName);
  }
}
