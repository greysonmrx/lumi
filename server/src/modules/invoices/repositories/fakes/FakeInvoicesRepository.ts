import { randomUUID } from "node:crypto";

import { ICreateInvoiceDTO } from "../../dtos/ICreateInvoiceDTO";

import { Invoice } from "../../entities/fakes/Invoice";

import { IInvoicesRepository } from "../IInvoicesRepository";

export class FakeInvoicesRepository implements IInvoicesRepository {
  public async create(data: ICreateInvoiceDTO): Promise<Invoice> {
    const invoice = new Invoice();

    const invoiceId = randomUUID();

    Object.assign(invoice, {
      id: invoiceId,
      dueDate: data.dueDate,
      items: data.items.map((item) => ({
        ...item,
        id: randomUUID(),
        invoiceId,
      })),
      customerId: randomUUID(),
      pdfId: randomUUID(),
      referenceMonth: data.referenceMonth,
      totalValue: data.totalValue,
      createdAt: new Date(),
    });

    return invoice;
  }
}
