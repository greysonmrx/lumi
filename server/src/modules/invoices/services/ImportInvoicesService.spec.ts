import {
  it,
  vi,
  describe,
  expect,
  afterEach,
  beforeEach,
  type MockInstance,
} from "vitest";

import { NotFoundError } from "@/shared/errors/NotFoundError";

import { FakeCustomersRepository } from "@/modules/customers/repositories/fakes/FakeCustomersRepository";
import { FakeFilesRepository } from "@/modules/files/repositories/fakes/FakeFilesRepository";

import { FakeFileManagerProvider } from "@/modules/files/providers/FileManagerProvider/implementations/FakeFileManagerProvider";

import { FakeInvoicesRepository } from "../repositories/fakes/FakeInvoicesRepository";
import { FakePdfReaderProvider } from "../providers/PdfReaderProvider/implementations/FakePdfReaderProvider";

import { ImportInvoicesService } from "./ImportInvoicesService";

describe("[SERVICES] - Import Invoices Service", () => {
  let fakeFilesRepository: FakeFilesRepository;
  let fakeInvoicesRepository: FakeInvoicesRepository;
  let fakeCustomersRepository: FakeCustomersRepository;
  let fakeFileManagerProvider: FakeFileManagerProvider;
  let fakePdfReaderProvider: FakePdfReaderProvider;

  let getCustomerNameSpy: MockInstance;
  let getCustomerNumberSpy: MockInstance;
  let getDueDateSpy: MockInstance;
  let getReferenceMonthSpy: MockInstance;
  let getTotalValueSpy: MockInstance;

  let importInvoices: ImportInvoicesService;

  const fileContent = "content" as any;
  const files = [
    { name: "invoice-1", size: 1000, content: fileContent },
    { name: "invoice-2", size: 2000, content: fileContent },
    { name: "invoice-3", size: 3000, content: fileContent },
  ];

  beforeEach(() => {
    fakeFilesRepository = new FakeFilesRepository();
    fakeInvoicesRepository = new FakeInvoicesRepository();
    fakeCustomersRepository = new FakeCustomersRepository();
    fakeFileManagerProvider = new FakeFileManagerProvider();
    fakePdfReaderProvider = new FakePdfReaderProvider();

    importInvoices = new ImportInvoicesService(
      fakeFilesRepository,
      fakeInvoicesRepository,
      fakeCustomersRepository,
      fakePdfReaderProvider,
      fakeFileManagerProvider
    );

    getCustomerNameSpy = vi.spyOn(importInvoices, "getCustomerName");
    getCustomerNumberSpy = vi.spyOn(importInvoices, "getCustomerNumber");
    getDueDateSpy = vi.spyOn(importInvoices, "getDueDate");
    getReferenceMonthSpy = vi.spyOn(importInvoices, "getReferenceMonth");
    getTotalValueSpy = vi.spyOn(importInvoices, "getTotalValue");
  });

  afterEach(() => {
    getCustomerNameSpy.mockClear();
    getCustomerNumberSpy.mockClear();
    getDueDateSpy.mockClear();
    getReferenceMonthSpy.mockClear();
    getTotalValueSpy.mockClear();
  });

  it("should be able to import invoices correctly", async () => {
    getCustomerNameSpy.mockReturnValue("Customer Name");
    getCustomerNumberSpy.mockReturnValue("123321");
    getDueDateSpy.mockReturnValue("2003-02-01");
    getReferenceMonthSpy.mockReturnValue("FEV/03");
    getTotalValueSpy.mockReturnValue(1000);

    const invoices = await importInvoices.perform({ files });

    expect(invoices).toHaveLength(3);

    invoices.forEach((invoice) => {
      expect(invoice).toMatchSnapshot({
        id: expect.any(String),
        createdAt: expect.any(Date),
        dueDate: expect.any(Date),
        pdfId: expect.any(String),
        customerId: expect.any(String),
        referenceMonth: expect.any(String),
        totalValue: expect.any(Number),
      });
    });
  });

  it("should not be able to import invoices if customer name is not found", async () => {
    getCustomerNumberSpy.mockReturnValue("123321");
    getDueDateSpy.mockReturnValue("2003-02-01");
    getReferenceMonthSpy.mockReturnValue("FEV/03");
    getTotalValueSpy.mockReturnValue(1000);

    await expect(importInvoices.perform({ files })).rejects.toThrow(
      NotFoundError
    );
  });

  it("should not be able to import invoices if customer number is not found", async () => {
    getCustomerNameSpy.mockReturnValue("Customer Name");
    getDueDateSpy.mockReturnValue("2003-02-01");
    getReferenceMonthSpy.mockReturnValue("FEV/03");
    getTotalValueSpy.mockReturnValue(1000);

    await expect(importInvoices.perform({ files })).rejects.toThrow(
      NotFoundError
    );
  });

  it("should not be able to import invoices if due date is not found", async () => {
    getCustomerNameSpy.mockReturnValue("Customer Name");
    getCustomerNumberSpy.mockReturnValue("123321");
    getReferenceMonthSpy.mockReturnValue("FEV/03");
    getTotalValueSpy.mockReturnValue(1000);

    await expect(importInvoices.perform({ files })).rejects.toThrow(
      NotFoundError
    );
  });

  it("should not be able to import invoices if reference month is not found", async () => {
    getCustomerNameSpy.mockReturnValue("Customer Name");
    getCustomerNumberSpy.mockReturnValue("123321");
    getDueDateSpy.mockReturnValue("2003-02-01");
    getTotalValueSpy.mockReturnValue(1000);

    await expect(importInvoices.perform({ files })).rejects.toThrow(
      NotFoundError
    );
  });

  it("should not be able to import invoices if total value is not found", async () => {
    getCustomerNameSpy.mockReturnValue("Customer Name");
    getCustomerNumberSpy.mockReturnValue("123321");
    getDueDateSpy.mockReturnValue("2003-02-01");
    getReferenceMonthSpy.mockReturnValue("FEV/03");

    await expect(importInvoices.perform({ files })).rejects.toThrow(
      NotFoundError
    );
  });
});
