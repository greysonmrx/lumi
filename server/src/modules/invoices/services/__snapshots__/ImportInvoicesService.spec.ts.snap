// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[SERVICES] - Import Invoices Service > should be able to import invoices correctly 1`] = `
{
  "createdAt": Any<Date>,
  "customer": undefined,
  "customerId": Any<String>,
  "dueDate": Any<Date>,
  "id": Any<String>,
  "items": [],
  "pdf": undefined,
  "pdfId": Any<String>,
  "referenceMonth": Any<String>,
  "totalValue": Any<Number>,
}
`;

exports[`[SERVICES] - Import Invoices Service > should be able to import invoices correctly 2`] = `
{
  "createdAt": Any<Date>,
  "customer": undefined,
  "customerId": Any<String>,
  "dueDate": Any<Date>,
  "id": Any<String>,
  "items": [],
  "pdf": undefined,
  "pdfId": Any<String>,
  "referenceMonth": Any<String>,
  "totalValue": Any<Number>,
}
`;

exports[`[SERVICES] - Import Invoices Service > should be able to import invoices correctly 3`] = `
{
  "createdAt": Any<Date>,
  "customer": undefined,
  "customerId": Any<String>,
  "dueDate": Any<Date>,
  "id": Any<String>,
  "items": [],
  "pdf": undefined,
  "pdfId": Any<String>,
  "referenceMonth": Any<String>,
  "totalValue": Any<Number>,
}
`;
